### ISSUE TITLE
Narrow typeof x === 'object' to Record<string | number | symbol, unknown>

### ISSUE THREAD

**Issue by MicahZoltu** [microsoft/TypeScript#38801]
## Search Terms
typeof object narrow record

## Suggestion
When narrowing a variable with `typeof x === 'object'` narrow to `Record<string | number | symbol, unknown> | null | unknown[]` instead of `object`.

## Use Cases
When testing to see if something is an object, you almost always are starting out with `unknown` or similarly near-top level type and trying to narrow down to a lower "useful" type.  When you use the `typeof x === 'object'` narrowing operation, currently TypeScript *over* narrows to the extremely constrained `{} | null` type.  This suggestion would change that behavior to narrow down to the much wider `Record<string | number | symbol, unknown> | null | unknown[]`, which the user can then choose to narrow more if they like, or they can choose to work with it as-is.

## Examples
```ts
declare const apple: unknown
if (typeof apple !== 'object') throw new Error() // actual: object | null; desired: Record<string|number|symbol, unknown> | null | unknown[]

// uncomment this line and comment out the lines above to see desired behavior
// declare const apple: Record<string|number|symbol, unknown> | null | unknown[]

if (apple === null) throw new Error() // actual: object; desired: Record<string|number|symbol, unknown> | unknown[]
if (Array.isArray(apple)) throw new Error() // actual: object; desired: Record<string|number|symbol, unknown>
for (const key in apple) {
    // Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{}'.
    //   No index signature with a parameter of type 'string' was found on type '{}'.
    apple[key] // actual: error; desired: unknown
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
I really doubt this wouldn't be an enormous breaking change, but we can look into it if there's more feedback

**Comment by MicahZoltu** [comment]
@RyanCavanaugh Do you have an example of code that this will break?  It seems like it is *loosening* the type checker, rather than tightening it, so any code that used to work should still work.

**Comment by RyanCavanaugh** [comment]
The great thing about users is that they write code you'd never expect.

Anyway one example would be
```ts
type IsArray<T> = T extends unknown[] ? true : false;
```
`IsArray<typeof something_narrowed_to_object>` is `false` today, the proposed change would make it `true | false`

**Comment by jonhue** [comment]
I think this change would be great! It would allow us to check the types of the operands to the `in` operator more accurately (#43210) without breaking existing code.

**Comment by 4086606** [comment]
Keep on seeing warnings related to #21732 and its rather irking..

**Comment by DetachHead** [comment]
> The great thing about users is that they write code you'd never expect.
> 
> Anyway one example would be
> 
> ```ts
> type IsArray<T> = T extends unknown[] ? true : false;
> ```
> 
> `IsArray<typeof something_narrowed_to_object>` is `false` today, the proposed change would make it `true | false`

i don't understand. `IsArray<Record<string | number | symbol, unknown>>` is still `false`, so i don't see the issue
https://www.typescriptlang.org/play?#code/C4TwDgpgBAkgzgQQE5IIYgDwBUB8UC8UWUEAHsBAHYAmcUArpQNaUD2A7pQNoC6UA-FGBJ60AFxQAZqgA2cCAG4AUEtCQoAMVasCsRCnQYAShADGrJNQxxhAS0oBzKAB8olegFsARhCQuocCDerDIANAzMbJw4OEA

**Comment by KotlinIsland** [comment]
@MicahZoltu You should update the title to: "Narrow `typeof x === 'object'` to `Record<string | number | symbol, unknown> | null | unknown[]`"

**Comment by SNDST00M** [comment]
Also shouldn't we be combining that `Record` with `null` anyway? As that's the literal defn of `typeof x === 'object'` 🤔

**Comment by k2snowman69** [comment]
Given that this would be an "enormous breaking change" can someone provide guidance of how we would go from a typeof x === "object" check to narrow it to a Record? Just some sample code like:

```
function assertStringToStringMap(tbd: unknown): Record<string, string> {
    if (tbd == null || !(typeof tbd === "object")) {
        throw new Error("tbd isn't an object, can't be a string to string map")
    }
    
    // Magic to check all keys are strings
    // More magic to check all keys are strings

    // Now TBD should be able to be a Record<string, string> so we can just return it
    return tbd;
}
```

But with the magic filled out?

**Comment by draymond63** [comment]
@k2snowman69 using a typeguard as suggested in [this reddit post](https://www.reddit.com/r/typescript/comments/ns6595/is_there_a_way_to_narrow_an_unknown_type_to/) probably works for most people's use cases.

```
function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}
```

**Comment by k2snowman69** [comment]
@draymond63 I think you missed my question... the question I was asking was what goes in that function to properly narrow an unknown to a Record. For example, I could also do this:

```
function isRecord(value: unknown): value is Record<string, unknown> {
    return true;
}
```

Which at that point I don't see the difference between a typeguard and a force cast `as Record<string, unknown>`. For now the benefit is until the "magic" is determined, at least the logic is in one place. However I'm more curious what that "magic" should actually be.

**Comment by enricobertolotti** [comment]
> function isRecord(value: unknown): value is Record<string, unknown> {
>   return typeof value === 'object' && value !== null;
> }

Yes thanks! But make sure also to make sure that the value is not an array (array is also of type 'object' in javascript)

```ts
function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
```
